akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala

```diff
@@ -567,19 +567,19 @@ class HttpExt(private val config: Config)(implicit val system: ActorSystem) exte
     val gatewayPromise = Promise[PoolGateway]()
     hostPoolCache.putIfAbsent(setup, gatewayPromise.future) match {
       case null ⇒ // only one thread can get here at a time
-        val whenShuttingDown = Promise[Done]()
+        // holds the reference to the fastFuture, so we can remove it on the shutdown started hook of the pool gateway
+        var fastFuture: Future[PoolGateway] = null
         val gateway =
-          try new PoolGateway(setup, whenShuttingDown)
+          try new PoolGateway(setup, () => hostPoolCache.remove(setup, fastFuture))
```

Comment:
I was thinking it might have to be volatile or such, but the variable is set from line 580-on and the `PoolGateway` hook is only executed when `poolGateway.shutdown()` is called. Which shouldn't happen in this time-frame (as far as I can tell) because nothing else holds a reference to the gateway instance. The `PoolInterfaceActor` can issue a `shutdown()`, but only after the idle timeout, which "shouldn't" occur before the variable has been assigned.

After thinking about it a little, it can be worked around by having the `PoolGateway` itself have a `val selfFuture = FastFuture.successful(this)`. Provided we change the `onShutdownStarted` hook to accept the `PoolGateway` instance as the first argument, we can use the `selfFuture` as the value in the `hostPoolCache`. That should make it easier to argue about.

Since the PR referenced in the conversation (https://github.com/akka/akka/pull/20190) changes a bunch of things around this code, it makes sense to only revisit this if the problem it's trying to solve still occurs after it's merged?